### PR TITLE
Add `eio` implementation

### PR DESCRIPTION
**Pull request by quernd**
Both unary and streaming, client and server are implemented. The API _mostly_ follows the existing API but we can take this opportunity to overhaul it.

### CODE REVIEW

**Comment by wokalski** [comment]
We'll iterate on it for our internal project and merge when ready.

File: examples/greeter-client-eio/dune
Comment (tmcgilchrist) [inline examples/greeter-client-eio/dune]:
```suggestion
 (libraries grpc grpc-eio ocaml-protoc eio_main greeter h2 h2-eio))
```
This should depend on `eio_main`

File: examples/greeter-server-eio/dune
Comment (tmcgilchrist) [inline examples/greeter-server-eio/dune]:
```suggestion
 (libraries grpc grpc-eio ocaml-protoc eio_main greeter h2 h2-eio))
```

File: examples/greeter-client-eio/greeter_client_eio.ml
Comment (tmcgilchrist) [inline examples/greeter-client-eio/greeter_client_eio.ml]:
```suggestion
    Eio.Promise.await (H2_eio.Client.shutdown connection);
```
This now returns a Promise that needs to be waited on.

**Review by tmcgilchrist — COMMENTED** [review]
This changes are now required to run the greeter example code.

**Comment by tmcgilchrist** [comment]
The last two commits on https://github.com/tmcgilchrist/ocaml-grpc/tree/grpc-eio provide the fixes and ports the route guide example to EIO.

File: dune-project
Comment (tmcgilchrist) [inline dune-project]:
```suggestion
  eio_main
```

**Comment by quernd** [comment]
Thanks @tmcgilchrist. I think the rebasing went well this time. CI is now all green except for the experimental lint workflows. I converted the greeter EIO examples to `ocaml-protoc-plugin` because the shared `greeter` library changed. Since you ported the routeguide tutorial to EIO, there are probably a few changes needed in the instructions. I only auto-generated the MDX code blocks. It's great that we already have a tutorial that uses EIO!

File: dune-project
Comment (tmcgilchrist) [inline dune-project:111]:
```suggestion
  (h2-eio (> 0.10.0))
```

File: lib/grpc-eio/connection.ml
Comment (tmcgilchrist) [inline lib/grpc-eio/connection.ml:21]: Should this be calling flush after each iteration?
I'm undecided about whether it should. The way it is now will batch up the writes and send them in a smaller number of network packets. Flushing after each `write_string` would send them immediately to the server, which might be preferable.

File: lib/grpc-eio/server.ml
Comment (tmcgilchrist) [inline lib/grpc-eio/server.ml:42]: This details the responses required for unsupported encodings https://github.com/grpc/grpc/blob/master/doc/compression.md#test-cases

File: dune-project
Comment (tmcgilchrist) [inline dune-project]:
```suggestion
  (eio_main (>= 0.10))
```
Add a lower bound on this? I am generally in favour of including them for libraries, plus ocaml-ci will check the project works with lower bounds.

File: examples/routeguide-tutorial.md
Comment (tmcgilchrist) [inline examples/routeguide-tutorial.md:1]: Undecided whether this should switch to EIO or stay with LWT. 
Either way, I think keep the lwt example code. The details of handling lwt_streams correctly is good to keep.

File: examples/greeter-client-eio/greeter_client_eio.ml
Comment (tmcgilchrist) [inline examples/greeter-client-eio/greeter_client_eio.ml]:
```suggestion
    let addr = `Tcp (Eio_unix.Net.Ipaddr.of_unix inet, port) in
```

File: examples/routeguide/src/client.ml
Comment (tmcgilchrist) [inline examples/routeguide/src/client.ml]:
```suggestion
  let addr = `Tcp (Eio_unix.Net.Ipaddr.of_unix inet, port) in
```

**Review by tmcgilchrist — APPROVED** [review]

File: dune-project
Comment (quernd) [reply dune-project]: Thanks! Agreed.

File: examples/routeguide-tutorial.md
Comment (quernd) [reply examples/routeguide-tutorial.md:1]: It's definitely better to have both examples, good call!

File: lib/grpc-eio/connection.ml
Comment (quernd) [reply lib/grpc-eio/connection.ml:21]: That's a good question. It could be preferable when aiming for low latency. I'm thinking maybe it should just be opt-in.

On this note, we ran some load tests but they were inconclusive because we stumbled upon a bug in Eio: https://github.com/ocaml-multicore/eio/issues/572
It has been fixed, so I hope we'll have better numbers soon.

We're also experimenting with a different interface for the reader. Lazy sequence might be a bit too opinionated, and they introduce a bit of overhead, so the alternative interface is similar to H2 with `on_msg`/`on_eof` callbacks, with a conversion to `Seq.t` as a convenience helper: https://github.com/dialohq/ocaml-grpc/tree/grpc-eio-noseq/lib/grpc-eio

File: lib/grpc-eio/server.ml
Comment (quernd) [reply lib/grpc-eio/server.ml:42]: Thanks! I'll take a look. By the way, with the Eio implementation and your Async server implementation, this code is now duplicated and could be factored out.

File: examples/greeter-server-eio/greeter_server_eio.ml
Comment (tmcgilchrist) [inline examples/greeter-server-eio/greeter_server_eio.ml:58]:
```suggestion
    {| dune exec -- examples/greeter-client-eio/greeter_client_eio.exe <your_name> |};
```

File: lib/grpc-eio/connection.ml
Comment (tmcgilchrist) [reply lib/grpc-eio/connection.ml:21]: Interesting experiment, it will be good to see some performance numbers from it. Keeping the `Seq.t` for a simple interface is a good idea for introducing the library and less performance sensitive applications. 

Opt-in sounds good, I'm happy to leave that till later on. There are a few other config pieces like auth and compression that will need a way to configure them.

**Review by tmcgilchrist — APPROVED** [review]

File: lib/grpc-eio/server.ml
Comment (tmcgilchrist) [reply lib/grpc-eio/server.ml:42]: I'll do another pass over the tutorial code after this is merged.

**Comment by tmcgilchrist** [comment]
Merging this to unblock other PRs.

**Pull request merged**
